pyscript.core/src/plugins/py-editor.js

```diff
@@ -0,0 +1,207 @@
+    outDiv.className = "py-editor-output";
+    outDiv.id = getID() + "-output";
+    return outDiv;
+};
+
+const makeBoxDiv = (listener) => {
+    const boxDiv = document.createElement("div");
+    boxDiv.className = "py-editor-box";
+
+    const editorDiv = makeEditorDiv(listener);
+    const outDiv = makeOutDiv();
+    boxDiv.append(editorDiv, outDiv);
+
+    return [boxDiv, outDiv];
+};
+
+const init = async (script) => {
+    const [
+        { basicSetup, EditorView },
+        { Compartment },
+        { python },
+        { indentUnit },
+        { keymap },
+        { defaultKeymap },
+    ] = await Promise.all([
+        // TODO: find a way to actually produce these bundles locally
+        import("https://cdn.jsdelivr.net/npm/codemirror@6.0.1/+esm"),
+        import("https://cdn.jsdelivr.net/npm/@codemirror/state@6.3.1/+esm"),
+        import(
+            "https://cdn.jsdelivr.net/npm/@codemirror/lang-python@6.1.3/+esm"
+        ),
+        import("https://cdn.jsdelivr.net/npm/@codemirror/language@6.9.2/+esm"),
+        import("https://cdn.jsdelivr.net/npm/@codemirror/view@6.22.0/+esm"),
+        import("https://cdn.jsdelivr.net/npm/@codemirror/commands@6.3.0/+esm"),
+    ]);
+
+    const selector = script.getAttribute("target");
+
+    let target;
+    if (selector) {
+        target =
+            document.getElementById(selector) ||
+            document.querySelector(selector);
+        if (!target) throw new Error(`Unknown target ${selector}`);
+    } else {
+        target = document.createElement("py-editor");
+        target.style.display = "block";
+        script.after(target);
+    }
+
+    if (!target.id) target.id = getID();
+    if (!target.hasAttribute("exec-id")) target.setAttribute("exec-id", 0);
+    if (!target.hasAttribute("root")) target.setAttribute("root", target.id);
+
+    // @see https://github.com/JeffersGlass/mkdocs-pyscript/blob/main/mkdocs_pyscript/js/makeblocks.js
+    const context = {
+        get pySrc() {
+            return editor.state.doc.toString();
+        },
+        get outDiv() {
+            return outDiv;
+        },
+    };
+
+    const listener = execute.bind(context);
+    const [boxDiv, outDiv] = makeBoxDiv(listener);
+
+    const inputChild = boxDiv.querySelector(".py-editor-input > div");
+    const parent = inputChild.attachShadow({ mode: "open" });
+    // avoid inheriting styles from the outer component
+    parent.innerHTML = `<style> :host { all: initial; }</style>`;
+
+    target.appendChild(boxDiv);
+
+    const doc = dedent(script.textContent).trim();
+
+    // preserve user indentation, if any
+    const indentation = /^(\s+)/m.test(doc) ? RegExp.$1 : "    ";
+
+    const editor = new EditorView({
+        extensions: [
+            indentUnit.of(indentation),
+            new Compartment().of(python()),
+            keymap.of([
+                ...defaultKeymap,
+                { key: "Ctrl-Enter", run: listener, preventDefault: true },
```

Comment:
TODO: These don't actually consistently work everywhere, with CodeMirror 😇 would be nice to figure out why.

Reply:
I think for Windows and Linux we're fine, there's the Mac Command+Enter to consider though ... as nobody presses Ctrl in there

Reply:
Done, I've just added Cmd-Enter too